Added infinite recursion tests

namesty opened the pull request:
Closes #149.

Detecting GraphQL recursive types is a challenge, as far as I know, with current tooling. So, by doing research, I found the following paper, on this exact problem, from a swedish university: http://liu.diva-portal.org/smash/get/diva2:1302887/FULLTEXT01.pdf.

I also found this repo: https://github.com/LiUGraphQL/graphql-schema-cycles-webapp. This repo's based on the previously mentioned paper and contains the javascript implementation we needed. However the code was ugly, it also contained an unnecessary (for our use case) graphicating function, and the dependencies were not handled properly.

So I took their code, organized it as a monorepo, migrated it to typescript (has still lots of `any`s as I did not want to lose too much typing figuring out all types), wrote tests for it, added a simple to use, and typed, interface and we are now using it as an external dependency: https://github.com/namesty/graphql-cycles-detector

I'd appreciate some guidance on how to properly license the package, add the README, and give proper credits to do a proper publish under the web3api domain.

I also think this could be a very useful and interesting contribution to the current GraphQL ecosystem and tooling.

namesty committed:
Added infinite recursion tests with temp package

namesty committed:
Updated package to detect cycles

namesty committed:
Merge branch 'prealpha-dev' into graphql-infinite-recursions

benefacto reviewed (commented):
LGTM! One suggestion, though: maybe we should use "cyclic" rather than "infinite" in the code. That'd be more in line with the the error we're displaying to the user and what's actually being tested for. We don't want cycles because they'd be infinite but we're testing for the cyclic references themselves, not infinity, as that might be difficult to test for :) .

dOrgJelli committed:
Merge branch 'prealpha-dev' into graphql-infinite-recursions

dOrgJelli commented:
Looks great @namesty! It looks like the graphify.ts file is throwing "Field - reference not defined" for the parse's sanity schema: https://github.com/Web3-API/monorepo/blob/prealpha/packages/test-cases/cases/parse/sanity/input.graphql

From what I can tell, it doesn't like the "importedMethod" function on the "TestImport_Mutation" type (I think). Maybe we need to give the package a "ignore" function which allows us to ignore types of certain names. For example, ignore everything by the name of "Query", "Mutation", "..._Query", and "..._Mutation".

namesty committed:
Merge remote-tracking branch 'origin/graphql-infinite-recursions' into graphql-infinite-recursions

namesty committed:
[WIP] edge cases and ignore type names

namesty committed:
Updated cycles detector to ignore Operation types - updated tests

namesty commented:
@dOrgJelli You were right. Also, it doesn't like custom directives with arguments either, since this edge case wasn't contemplated by the original authors.

So I updated the package to accept a parameter to ignore types, and to remove custom dependencies for the circular definition evaluation and bumped the version number. I then updated the usage in the web3api monorepo to search for all type definitions that contain "Mutation", "Query" or "Subscription" and pass them to the validator as types to be ignored.

This worked, and detected circular dependencies in the bind and parse sanity checks, so I updated those and added more test cases to the infinite recursions test section.

Finally, I updated the naming as per @benefacto's recommendation, `infinite recursion` -> `circular definition`

dOrgJelli commented on packages/cli/src/__tests__/project/test/test/package.json:
```diff
@@ -0,0 +1,5 @@
+
```
Can this file be removed? Is it used anywhere?

dOrgJelli replied:
Removed in my updates

dOrgJelli commented on packages/schema/parse/src/validate/types.ts:
```diff
@@ -149,3 +150,33 @@ export function propertyTypes(astNode: DocumentNode): void {
+    enter: {
+      ObjectTypeDefinition: (node) => {
+        const isOperationType = operationTypeNames.some((name) =>
+          node.name.value.includes(name)
```
This doesn't work for types like `type FooQueryOptions { ... }`. We instead want to look for `name === "Query" | "Mutation" | "Subscription"`, and also `name.endsWith("_Query" | "_Mutation" | "_Subscription")`

dOrgJelli replied:
Fixed in my updates

dOrgJelli committed:
update graphql version + remove unused project + add back in valid circular relationship

dOrgJelli commented on packages/schema/parse/src/__tests__/validate-types.spec.ts:
```diff
@@ -118,6 +118,95 @@ type Bar {
+type B {
+  prop: A!
+}
+`
```
Let's add a test for:
```
type A {
  b: B!
}

type B {
  a: A
}
```
to make sure this doesn't throw an error, as it's a valid relationship.

namesty committed:
Allow circular reference on nullable fields

namesty commented:
I updated the package to contemplate an option to allow for circular references on nullable fields, as it was requested above. I made this on the package side, and a configurable option as that is our Web3API's use case, but probably another project that also uses the package would want otherwise.

Anyways, after the update all tests are passing and I added the requested test case :D

namesty committed:
Merge branch 'prealpha-dev' into graphql-infinite-recursions

dOrgJelli committed:
fix bind test case

dOrgJelli committed:
lint fix

dOrgJelli committed:
graphql version bump

dOrgJelli committed:
reset yarn.lock
